File: src/HitCalc.java
import java.text.DecimalFormat;
import java.util.*;

/**
 * The main interface of the application.
 */
public final class HitCalc {

    /**
     * The main method launching the application.
     *
     */
    public static final void main(String[] args) {
        final Scanner scanner = new Scanner(System.in);
        final DecimalFormat df = new DecimalFormat("#.##");

        double ac = 0;
        double mod = 0;

        System.out.println("Numbers only, 0 to exit");

        while (true) {
            System.out.print("\nEnemy's AC: ");
            try {
                ac = scanner.nextDouble();
            } catch (InputMismatchException ime) {
                System.out.println("NUMBERS ONLY YOU CUNT");
                System.exit(1);
            }

            if (ac == 0) {
                System.exit(0);
            }

            System.out.print("Mod: ");
            try {
                mod = scanner.nextDouble();
            } catch (InputMismatchException ime) {
                System.out.println("NUMBERS ONLY YOU CUNT");
                System.exit(1);
            }

            final double probHit = (20 - ac + mod + 1) / 20;
            final double probMiss = (ac - mod - 1) / 20;
            final double resultAdv = 1 - (probMiss * probMiss);
            final double resultDis = probHit * probHit;

            if (probHit <= 0) {
                System.out.println("Impossible to hit, unless you roll a 20");
            } else if (probHit >= 1) {
                System.out.println("Guaranteed hit, unless you roll a 1");
            } else {
                System.out.printf("Normal hit rate: %s\nAdvantage: %s - Disadvantage: %s\n",
                        df.format(probHit), df.format(resultAdv), df.format(resultDis));
            }
        }
    }
}

File: c/HitCalc.c
#include <stdio.h>
#include <stdlib.h>

int main(int argc, const char * argv[]) {
    int ac = atoi(argv[1]);
    int mod = atoi(argv[2]);

    double probHit = 20 - ac + mod + 1;
    probHit /= 20;  //fix to prevent integer division
    double probMiss = ac - mod - 1;
    probMiss /= 20;

    double resultAdv = 1 - (probMiss * probMiss);
    double resultDis = probHit * probHit;

    if (probHit <= 0) {
        printf("Impossible to hit, unless you roll a 20\n");
    } else if (probHit >= 1) {
        printf("Guaranteed hit, unless you roll a 1\n");
    } else {
        printf("Enemy's AC: %d - Mod: %d\nNormal hit rate: %g\nAdvantage: %g - Disadvantage: %g\n", 
            ac, mod, probHit, resultAdv, resultDis);
    }
}